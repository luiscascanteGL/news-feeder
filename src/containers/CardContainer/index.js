import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from "react-redux";
import { fetchArticles } from "../../actions";
import CardComponent from "../../components/Card";

import "./index.scss";

export const CardContainer = ({ isFetching, articles, fetchArticles }) => {

  const [searchText, setSearchText] = useState("");
  const [selectedSource, setSelectedSource] = useState("");
  const [filteredArticles, setFilteredArticles] = useState([]);
  let page = 0;

  useEffect(() => {
    fetchArticles(page);
    window.addEventListener("scroll", handleScroll);
  }, []);

  useEffect(() => {
    setFilteredArticles(filterArticles());
  }, [searchText, selectedSource, articles]);

  const handleTextChange = (event) => {
    setSearchText(event.target.value);
  }

  const handleSourceChange = (event) => {
    setSelectedSource(event.target.value);
  }

  const handleSubmit = (event) => {
    event.preventDefault();
  }

  const handleScroll = async () => {
    const windowHeight = "innerHeight" in window ? window.innerHeight : document.documentElement.offsetHeight;
    const body = document.body;
    const html = document.documentElement;
    const docHeight = Math.max(body.scrollHeight, body.offsetHeight, html.clientHeight, html.scrollHeight, html.offsetHeight);
    const windowBottom = Math.round(windowHeight + window.pageYOffset);
    if (windowBottom >= docHeight) {
      page = page + 1;
      await fetchArticles(page);
    }
  }

  const checkSourceMatch = ({ source }) => {
    try {
      return source.hasOwnProperty("name") && source.name === selectedSource;
    }
    catch (error) {
      console.log(error);
    }
  }

  const checkTextMatch = ({ title = "", description = "", author = "", content = "" }) => {
    try {
      return title.includes(searchText) || description.includes(searchText) || author.includes(searchText) || content.includes(searchText);
    }
    catch (error) {
      console.log(error);
    }
  }

  const sources = articles.map(({ source }) => {
    return (<option key={`${source.name}`} value={source.name}>{source.name}</option>);
  })


  const filterArticles = () => {
    if (searchText !== "" && selectedSource !== "") {
      try {
        return articles.filter((article) => checkTextMatch(article) && checkSourceMatch(article));
      }
      catch (error) {
        console.log(error);
        return articles;
      }
    }
    else if (searchText !== "") {
      try {
        return articles.filter((article) => checkTextMatch(article));
      }
      catch (error) {
        console.log(error);
        return articles;
      }
    }
    else if (selectedSource !== "") {
      return articles.filter((article) => checkSourceMatch(article));
    }
    else {
      return articles;
    }
  }



  return (
    <div className="container">
      <form className="form--container" onSubmit={handleSubmit}>
        <div className="form--group">
          <label htmlFor="sources">Filter by</label>
          <select name="sources" id="sources" onChange={handleSourceChange}><option value="">All</option>{sources}</select>
        </div>
        <div className="form--group">
          <label htmlFor="keyword">Search</label>
          <input name="keyword" id="keyword" placeholder="Type a keyword..." type="text" value={searchText} onChange={handleTextChange} />
        </div>
      </form>
      <div className="card--list">
        {filteredArticles.map((article) => <CardComponent key={`${article.title}`} article={article} />)}
        <div style={{ width: '400px', padding: '0 2% 0 2%' }} />
      </div>
      {isFetching ? "Loading..." : null}
    </div>
  )
}

const mapStateToProps = state => {
  return {
    isFetching: state.newsReducer.isFetching,
    articles: state.newsReducer.articles
  }
}
const mapDispatchToProps = dispatch => {
  return {
    fetchArticles: (page) => dispatch(fetchArticles(page))
  }
}

CardContainer.propTypes = {
  isFetching: PropTypes.bool,
  articles: PropTypes.array,
  fetchArticles: PropTypes.func
}

export default connect(mapStateToProps, mapDispatchToProps)(CardContainer);


