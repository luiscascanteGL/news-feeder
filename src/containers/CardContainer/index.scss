@import "../../styles/variables.scss";

.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: #e7e7e7;
}

.form--container {
  width: 90%;
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  background-color: #fff;

  font-family: $base-family;
  font-size: $base-size;
 
  @media (min-width: $screen-md-min){
    flex-direction: row;
    align-items: center;
  }

  .form--group {
    display: flex;

    label {
      padding: 2px 5px 2px 5px;
    }

    input, select {
      margin-left: 5px;
      border: 2px solid grey;
      padding: 2px 5px 2px 5px;
      border-radius: 5px;
    }
  }
}

.card--list{
  margin: 2% 5% 2% 5%;
  padding: 10px 0 10px 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 90%;
  background-color: #fff;

  @media (min-width: $screen-md-min) {
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-evenly;
    align-items: center;
  }
}