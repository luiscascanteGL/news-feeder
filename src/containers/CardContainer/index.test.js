import React from 'react';
import { mount } from 'enzyme';
import CardContainer from "./index";

import { Provider } from "react-redux";
import { createStore, applyMiddleware } from "redux";
import reducer from "../../reducers"
import thunk from "redux-thunk"

const store = createStore(reducer, applyMiddleware(thunk))


describe('Card Container', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = mount(<Provider store={store}><CardContainer /></Provider>)
  });

  it('should render correctly', () => expect(wrapper).toMatchSnapshot());

  it('Should render a <div/>', () => {
    expect(wrapper.first('div').length).toEqual(1);
  })

  it('Should render a <form/>', () => {
    expect(wrapper.find('form').length).toEqual(1);
  })

  it('Should render a <input/>', () => {
    expect(wrapper.find('input').length).toEqual(1);
  })

  it('Should render a <select/>', () => {
    expect(wrapper.find('select').length).toEqual(1);
  })

  it('Should render two <label/>', () => {
    expect(wrapper.find('label').length).toEqual(2);
  })
})