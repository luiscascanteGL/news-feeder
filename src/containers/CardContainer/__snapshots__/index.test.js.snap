// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Container should render correctly 1`] = `ReactWrapper {}`;
